### ISSUE TITLE
Tests of yom hazikaron and yom haatsmaut fail sometimes

### ISSUE THREAD

**Issue by tsvi** [royi1000/py-libhdate#36]
Travis lately reports some failures on unittests for yom hazikaron and yom haatsmaut.

Build 207.4
```
 TestSpecialDays.test_get_holidays_moving[possible_dates5-years5-yom_haatzmaut] 
self = <tests.test_hdate.TestSpecialDays object at 0x7fbffa82cb00>
possible_dates` = [(3, 8), (4, 8), (5, 8), (6, 8)], years = (5764, 6500)
holiday = 'yom_haatzmaut'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:240: AssertionError
```

Build 208.1
```
 TestSpecialDays.test_get_holidays_moving[possible_dates5-years5-yom_haatzmaut] 
self = <tests.test_hdate.TestSpecialDays object at 0x7f12bdef2590>
possible_dates = [(3, 8), (4, 8), (5, 8), (6, 8)], years = (5764, 6500)
holiday = 'yom_haatzmaut'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:240: AssertionError
```

Build 208.2
```
 TestSpecialDays.test_get_holidays_moving[possible_dates5-years5-yom_haatzmaut] 
self = <tests.test_hdate.TestSpecialDays object at 0x7f386ec91c50>
possible_dates = [(3, 8), (4, 8), (5, 8), (6, 8)], years = (5764, 6500)
holiday = 'yom_haatzmaut'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:240: AssertionError
```

[Build 210.4] (https://travis-ci.org/tsvi/py-libhdate/jobs/467473473)

```
 TestSpecialDays.test_get_holidays_moving[possible_dates5-years5-yom_haatzmaut] 
self = <tests.test_hdate.TestSpecialDays object at 0x7efd75bb2748>
possible_dates = [(3, 8), (4, 8), (5, 8), (6, 8)], years = (5764, 6500)
holiday = 'yom_haatzmaut'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:241: AssertionError
```

[Build 212.2] (https://travis-ci.org/tsvi/py-libhdate/jobs/467473777)

```
 TestSpecialDays.test_get_holidays_moving[possible_dates5-years5-yom_haatzmaut] 
self = <tests.test_hdate.TestSpecialDays object at 0x7f9cc822c5c0>
possible_dates = [(3, 8), (4, 8), (5, 8), (6, 8)], years = (5764, 6500)
holiday = 'yom_haatzmaut'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:241: AssertionError
----------------------------- Captured stdout call -----------------------------
Testing yom_haatzmaut for 6379
date  Wednesday 5 Iyyar 6379 20 in the Omer Yom HaAtzma'ut  matched
checking  Monday 3 Iyyar 6379 18 in the Omer  doesn't match
checking  Tuesday 4 Iyyar 6379 19 in the Omer Yom HaZikaron  doesn't match
checking  Thursday 6 Iyyar 6379 21 in the Omer  doesn't match
Testing yom_haatzmaut for 5733
 TestSpecialDays.test_get_holidays_moving[possible_dates7-years7-yom_hazikaron] 
self = <tests.test_hdate.TestSpecialDays object at 0x7f9cc7d24080>
possible_dates = [(2, 8), (3, 8), (4, 8), (5, 8)], years = (5764, 6500)
holiday = 'yom_hazikaron'
    @pytest.mark.parametrize('possible_dates, years, holiday',
                             MOVING_HOLIDAYS)
    def test_get_holidays_moving(self, possible_dates, years, holiday):
        found_matching_holiday = False
        year = random.randint(*years)
    
        print("Testing " + holiday + " for " + str(year))
    
        for date in possible_dates:
            date_under_test = HDate(hebrew=False)
            date_under_test.hdate = HebrewDate(year, date[1], date[0])
            if date_under_test.holiday_name == holiday:
                print("date ", date_under_test, " matched")
                for other in possible_dates:
                    if other != date:
                        other_date = HDate(hebrew=False)
                        other_date.hdate = HebrewDate(year, other[1], other[0])
                        print("checking ", other_date, " doesn't match")
                        assert other_date.holiday_name != holiday
                found_matching_holiday = True
                assert date_under_test.is_holiday
    
        assert found_matching_holiday
    
        # Test holiday == 0 before 'since'
        # In case of yom hazikaron and yom ha'atsmaut don't test for the
        # case of 0 between 5708 and 5764
        if years[0] != 5000:
            if years[0] == 5764 and holiday in [17, 25]:
                return
            year = random.randint(5000, years[0] - 1)
            print("Testing " + holiday + " for " + str(year))
            for date in possible_dates:
                date_under_test = HDate()
                date_under_test.hdate = HebrewDate(year, date[1], date[0])
>               assert date_under_test.holiday_name == ""
E               AssertionError: assert 'yom_hazikaron' == ''
E                 - yom_hazikaron
tests/test_hdate.py:241: AssertionError
----------------------------- Captured stdout call -----------------------------
Testing yom_hazikaron for 6404
date  Friday 4 Iyyar 6404 19 in the Omer Yom HaZikaron  matched
checking  Wednesday 2 Iyyar 6404 17 in the Omer  doesn't match
checking  Thursday 3 Iyyar 6404 18 in the Omer  doesn't match
checking  Saturday 5 Iyyar 6404 20 in the Omer Yom HaAtzma'ut  doesn't match
Testing yom_hazikaron for 5755
```

**Comment by arigilder** [comment]
I fixed this with #37.

**Comment by tsvi** [comment]
I don't think so.

This part of the test is not supposed to run. The idea was to check that holidays that have a start year shouldn't return the holiday if it's before the given year. So you don't get Yom Haatsmaut in 1850.

So in the bracket of `(5764, 6500)`, we shouldn't run this test since during the years 5708-5764 there was yom ha'atsmaut but the rules as to when it fell were a bit different.

That's why we have `if years[0] == 5764 and holiday in [17, 25]: return`

I split the test in my latest commit (ae5bebf398), and I hope at some time to start adding docstrings and simplify the tests, as it took me to long to figure out what I was doing there.

**Comment by tsvi** [comment]
Keeping this open until I have seen enough tox runs where it doesn't happen.

**Comment by arigilder** [comment]
holiday in [17, 25] was a meaningless statement since holiday is a string (holiday name). I fixed that, and that return statement is now hit for the (5764, 6500) case, thus the test is not run. I specifically tested this, and ran tox a whole bunch of times to ensure the random hdate did not it. I also manually verified it by supplying a integer year (5804) which I knew to cause a failure. After the fix, the test passed.
